basin3d/core/catalog.py

```diff
@@ -479,170 +625,200 @@ def find_attribute_mappings(self, datasource_id: str = None, attr_type: str = No
+                    f'"{datasource_id}", attribute type = "{attr_type}", {vocab_source_type} '
+                    f'vocabularies: {attr_vocab and ",".join(attr_vocab) or None}.')
+            elif attr_vocab and len(attr_mappings) != len(attr_vocab):
+                # Find missing vocab in attr_mappings
                 if from_basin3d:
-                    results = self.in_memory_db_attr.search((query.basin3d_vocab.test(is_in)) & (query.datasource_id == datasource_id))
+                    not_found_attr_vocab = [vocab for vocab in attr_vocab if
+                                            vocab not in [am.basin3d_vocab for am in attr_mappings]]
                 else:
-                    results = self.in_memory_db_attr.search((query.datasource_vocab.test(is_in)) & (query.datasource_id == datasource_id))
+                    not_found_attr_vocab = [vocab for vocab in attr_vocab if
+                                            vocab not in [am.datasource_vocab for am in attr_mappings]]
+                logger.warning(
+                    f'No attribute mappings found for the following {vocab_source_type} '
+                    f'vocabularies: {",".join(not_found_attr_vocab)}. Note: specified datasource id = {datasource_id} and attribute type = {attr_type}.')
+        except SQLAlchemyError as e:
+            if not isinstance(e, NoResultFound):
+                raise e
 
-            # Finally, all parameters specified:
-            # Convert from BASIN-3D to DataSource variable name
-            elif from_basin3d:
-                results = self.in_memory_db_attr.search(
-                    (query.basin3d_vocab.test(is_in)) & (query.datasource_id == datasource_id) & (query.attr_type.search(attr_type)))
+        for attr_mapping in attr_mappings:
+            value = self._convert_attribute_mapping(attr_mapping)
+            if value:
+                yield value
 
-            # Convert from DataSource variable name to BASIN-3D
-            else:
-                results = self.in_memory_db_attr.search(
-                    (query.datasource_vocab.test(is_in)) & (query.datasource_id == datasource_id) & (query.attr_type.search(attr_type)))
+        session.close()
 
-            # Yield the results
-            attr_map: Optional[AttributeMapping]
-            for r in results:
-                attr_map = self._get_attribute_mapping(**r)
+    def _init_catalog(self, **kwargs):
+        """
+        Initialize the catalog database
+        """
+        if not self.is_initialized():
+            session = sqlalchemy_models.Session()
 
-                # if AttributeMapping is not found: THIS SHOULD NEVER HAPPEN.
-                if attr_map is None:
-                    logger.error('AttributeMapping not found given database search results. THIS SHOULD NEVER HAPPPEN.')
-                    continue
+            from basin3d.core.plugin import PluginMount
+            for name, plugin in PluginMount.plugins.items():
```

Comment:
My IDE is complaining that the `plugin` type is a `str `instead of `DataSourcePluginPoint`. Therefore in the lines below that call the method `plugin.get_meta()` is not valid because it thinks `plugin` is a `str`.

In the plugin.py module, the `PluginMount.plugins` element (line 224) is defined as a `Dict[DataSourcePluginPoint, str]`. So it seems like potentially, the for loop here should be `for plugin, name in PluginMount.plugins.items()`.

BUT none of our tests are failing which means the catalog is being initialized correctly. Not sure what is going on.

Reply:
hmm.  mypy and flake8 are not breaking on this so let's leave it.

Reply:
Making this a new issue.